File: java/src/main/java/it/giannibombelli/wsc2026/booking/domain/policies/BookingRefundRequestPolicy.java
package it.giannibombelli.wsc2026.booking.domain.policies;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.booking.domain.events.BookingResultEvents;
import it.giannibombelli.wsc2026.common.domain.model.Policy;
import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.payment.application.commands.RefundTransaction;
import it.giannibombelli.wsc2026.payment.domain.payment.Payment;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentRepository;

import static java.util.Objects.requireNonNull;



public class BookingRefundRequestPolicy implements Policy<BookingResultEvents.BookingRefused, RefundTransaction> {

    private final PaymentRepository paymentRepository;

    public BookingRefundRequestPolicy(PaymentRepository paymentRepository) {
        this.paymentRepository = requireNonNull(paymentRepository);
    }

    @Override
    public RefundTransaction evaluate(BookingResultEvents.BookingRefused event) {
        Require.requireArgument(event, "Booking refused event");

        ClientReference clientReference = new ClientReference(event.aggregateId().value().toString());
        Payment payment = paymentRepository.findByClientReference(clientReference)
            .orElseThrow(() -> new IllegalStateException("Payment not found for booking: " + clientReference.value()));

        return new RefundTransaction(payment.id(), event.amount());
    }
}

File: typescript/src/booking/domain/policies/bookingRefundRequestPolicy.ts
// Policy che traduce un BookingRefused in una richiesta di rimborso per il Payment BC.


import type { Policy } from '@/common/domain/model/policy.js';
import { ClientReference } from '@/common/domain/primitive/clientReference.js';
import { refundTransaction, type RefundTransaction } from '@/payment/application/commands/refundTransaction.js';
import type { PaymentRepository } from '@/payment/domain/ports/paymentRepository.js';
import { BookingRefused } from '../events/bookingResultEvents.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class BookingRefundRequestPolicy implements Policy<BookingRefused, RefundTransaction> {
  private readonly _paymentRepository: PaymentRepository;

  constructor(paymentRepository: PaymentRepository) {
    requireDependency(paymentRepository, "paymentRepository");
    this._paymentRepository = paymentRepository;
  }

  evaluate(event: BookingRefused): RefundTransaction {
    requireArgument(event, 'Booking refused event');

    const clientReference = new ClientReference(event.aggregateId.value.value);
    const payment = this._paymentRepository.findByClientReference(clientReference);
    if (payment === null) {
      throw new Error(`Payment not found for booking: ${clientReference.value}`);
    }

    return refundTransaction(payment.id(), event.amount);
  }
}
